Source/JavaScriptCore/runtime/JSONObject.cpp

```diff
@@ -627,6 +634,463 @@ bool Stringifier::Holder::appendNextProperty(Stringifier& stringifier, StringBui
+        return !function.isUndefined();
+    if (object.noSideEffectMayHaveNonIndexProperty(m_vm, m_vm.propertyNames->toJSON)) {
+        // Getting the property value so we can cache it could cause side effects; instead return true without caching anything.
+        return true;
+    }
+    // Cache this so we can answer false next time without redoing the noSideEffectMayHaveNonIndexProperty work.
+    PropertySlot slot { &object, PropertySlot::InternalMethodType::Get };
+    object.structure()->cacheSpecialProperty(&m_globalObject, m_vm, jsUndefined(), CachedSpecialPropertyKey::ToJSON, slot);
+    return false;
+}
+
+void FastStringifier::append(JSValue value)
+{
+    if (value.isNull()) {
+        if (remainingCapacity() < 4) {
+            recordBufferFull();
+            return;
+        }
+        m_buffer[m_length] = 'n';
+        m_buffer[m_length + 1] = 'u';
+        m_buffer[m_length + 2] = 'l';
+        m_buffer[m_length + 3] = 'l';
+        m_length += 4;
+        return;
+    }
+
+    if (value.isTrue()) {
+        if (remainingCapacity() < 4) {
+            recordBufferFull();
+            return;
+        }
+        m_buffer[m_length] = 't';
+        m_buffer[m_length + 1] = 'r';
+        m_buffer[m_length + 2] = 'u';
+        m_buffer[m_length + 3] = 'e';
+        m_length += 4;
+        return;
+    }
+
+    if (value.isFalse()) {
+        if (remainingCapacity() < 5) {
+            recordBufferFull();
+            return;
+        }
+        m_buffer[m_length] = 'f';
+        m_buffer[m_length + 1] = 'a';
+        m_buffer[m_length + 2] = 'l';
+        m_buffer[m_length + 3] = 's';
+        m_buffer[m_length + 4] = 'e';
+        m_length += 5;
+        return;
+    }
+
+    if (value.isInt32()) {
+        auto number = value.asInt32();
+        auto length = lengthOfIntegerAsString(number);
+        if (remainingCapacity() < length) {
+            recordBufferFull();
+            return;
+        }
+        writeIntegerToBuffer(number, &m_buffer[m_length]);
+        m_length += length;
+        return;
+    }
+
+    if (!value.isCell()) {
+        recordFailure("value type"_s);
+        return;
+    }
+    auto& cell = *value.asCell();
+
+    switch (cell.type()) {
+    case StringType: {
+        auto& string = asString(&cell)->tryGetValue();
+        if (string.isNull()) {
+            recordFailure("String::tryGetValue"_s);
+            return;
+        }
+        if (!string.is8Bit()) {
+            recordFailure("16-bit string"_s);
+            return;
+        }
+        auto stringLength = string.length();
+        if (remainingCapacity() < 1 + stringLength + 1) {
+            recordBufferFull();
+            return;
+        }
+        m_buffer[m_length] = '"';
+        auto* characters = string.characters8();
+        for (unsigned i = 0; i < stringLength; ++i) {
+            auto character = characters[i];
+            if (WTF::escapedFormsForJSON[character]) {
```

Comment:
To clarify, I meant that the expectation that the fast path will succeed most of the time implies that the fail out cases are unlikely.  Hence, it may be profitable to apply `UNLIKELY` to the fail out conditions.

Reply:
I won’t add the double case just because of your question "are we sure". But I am happy to return and add it if there is any evidence later. A trivial incremental step.

Reply:
I would be happy to add `UNLIKELY` if we have even a tiny bit of evidence it makes the code faster. I won’t do that experiment this time, but would be totally willing to follow up with that change.